### PR TITLE
Multi row inserts

### PR DESCRIPTION
**Pull request by hariso**

### CODE REVIEW

**Comment by ahmeroxa** [comment]
I think that although this does the job, I don't think it's quite aligned with the initial design. My understanding is that `Dialects` are there to capture the nuances of individual database types.

I imagined we'd add multi-row support to the `GenericDatabaseDialect` and then introduce a Redshift specific one that removes support for `insert` and defaults to `multi`, as well as rolls in the explicit mapping of `string` to `varchar`.

**Comment by hariso** [comment]
> I think that although this does the job, I don't think it's quite aligned with the initial design. My understanding is that `Dialects` are there to capture the nuances of individual database types.
> 
> I imagined we'd add multi-row support to the `GenericDatabaseDialect`

I wasn't comfortable adding multi-row support to `GenericDatabaseDialect`, since I wasn't able to actually confirm that all databases support it. I found  that for example, PostgreSQL didn't support this until 8.2 But if we can, for example, assume that it's only the minority of DBs (and versions) which do not support it, then that could be a reason for putting multi-row inserts into `GenericDatabaseDialect`.

> and then introduce a Redshift specific one that removes support for `insert` and defaults to `multi`, 
I definitely see advantages in that. The code would be faster by default. 

I feel, though, that a driver/DB dialect should reflect all of the functionality available in a DB. So if the DB itself gives both option (single row inserts and multi-row inserts), the driver should expose both of them too.

Maybe, as a middle ground, we could change `JdbcSinkConfig#INSERT_MODE_DEFAULT` to be `multi`?

> as well as rolls in the explicit mapping of `string` to `varchar`.
I planned doing that in another PR, to that their scope is smaller and it's easier to review. But I can definitely add it in this PR too.

**Comment by ahmeroxa** [comment]
> > I think that although this does the job, I don't think it's quite aligned with the initial design. My understanding is that `Dialects` are there to capture the nuances of individual database types.
> > I imagined we'd add multi-row support to the `GenericDatabaseDialect`
> 
> I wasn't comfortable adding multi-row support to `GenericDatabaseDialect`, since I wasn't able to actually confirm that all databases support it. I found that for example, PostgreSQL didn't support this until 8.2 But if we can, for example, assume that it's only the minority of DBs (and versions) which do not support it, then that could be a reason for putting multi-row inserts into `GenericDatabaseDialect`.
> 
> > and then introduce a Redshift specific one that removes support for `insert` and defaults to `multi`,
> > I definitely see advantages in that. The code would be faster by default.
> 
> I feel, though, that a driver/DB dialect should reflect all of the functionality available in a DB. So if the DB itself gives both option (single row inserts and multi-row inserts), the driver should expose both of them too.
> 
> Maybe, as a middle ground, we could change `JdbcSinkConfig#INSERT_MODE_DEFAULT` to be `multi`?
> 
> > as well as rolls in the explicit mapping of `string` to `varchar`.
> 
> I planned doing that in another PR, to that their scope is smaller and it's easier to review. But I can definitely add it in this PR too.

Ok that's fair.

I think we can set the default to `multi` for the Redshift dialect, but probably shouldn't change the default generally as insert performance is generally good on all other databases.

File: src/main/java/io/aiven/connect/jdbc/dialect/MultiInsertDialect.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                                            final Collection<ColumnId> keyColumns,
+                                            final Collection<ColumnId> nonKeyColumns) {
+
+        if (records < 1) {
+            throw new IllegalArgumentException("number of records must be a positive number, but got: " + records);
+        }
+        if (isEmpty(keyColumns) && isEmpty(nonKeyColumns)) {
+            throw new IllegalArgumentException("no columns specified");
+        }
+        requireNonNull(table, "table mustn't be null");
```
Comment (ahmeroxa) [inline src/main/java/io/aiven/connect/jdbc/dialect/MultiInsertDialect.java:57]:
```suggestion
        requireNonNull(table, "table must not be null");
```

**Review by ahmeroxa — APPROVED** [review]
LGTM 👍

**Comment by ahmeroxa** [comment]
@hariso what's left to do with this?

**Comment by hariso** [comment]
@ahmeroxa 

> @hariso what's left to do with this?

@lyuboxa wanted to review the PR too. On my  end, I still need to do end-to-end testing, but I was having some problems with my local setup and the CLI connecting to it. I made some progress on that, but I'll have a call with Diana to try getting that done today.

**Comment by hariso** [comment]
I finally got some test results, and they look promising. 

Current version:     25 437 records in 12.5 minutes.
Improved version:  50 000 records in cca 25 seconds

Roughly, the improvement is 60x. 

@ahmeroxa @lyuboxa

**Pull request merged**